src/landscape.cpp

```diff
@@ -994,6 +994,83 @@ static bool MakeLake(TileIndex tile, void *user_data)
 	return false;
 }
 
+/**
+ * Check whether a river could (logically) flow into a lock.
+ * @param tile the middle tile of a lock.
+ * @param recursive whether the function is being called recursively.
+ * @return true iff the water can be flowing into a lock.
+ */
+static bool FlowsLock(TileIndex tile, bool recursive = false)
+{
+	DiagDirection dir = GetInclinedSlopeDirection(GetTileSlope(tile));
+	if (dir == INVALID_DIAGDIR) return false;
+
+	int delta_mid = TileOffsByDiagDir(dir);
+	if (!IsTileFlat(tile + delta_mid)) return false;
+	if (!IsTileFlat(tile - delta_mid)) return false;
+
+	if (!recursive) return true;
+
+	DiagDirection dir_rot = ChangeDiagDir(dir, DIAGDIRDIFF_90RIGHT);
+	int delta = TileOffsByDiagDir(dir_rot);
+
+	for (int m = -1; m <= 1; m += 2) {
+		TileIndex t_dm = tile + m * delta_mid;
+		if (IsValidTile(t_dm)) {
+			if (DistanceFromEdgeDir(t_dm, dir) == 0 || DistanceFromEdgeDir(t_dm, ReverseDiagDir(dir)) == 0) return false;
+
+			for (int d = -1; d <= 1; d += 2) {
+				if (IsValidTile(t_dm + d * delta)) {
+					if (FlowsLockGivenDiagDir(t_dm + d * delta, dir_rot, true)) return false;
+
+					if (IsValidTile(t_dm + d * 2 * delta)) {
+						if (FlowsLockGivenDiagDir(t_dm + d * 2 * delta, dir_rot, true)) return false;
+					}
+				}
+			}
+
+			TileIndex t_2dm = t_dm + m * delta_mid;
```

Comment:
Variable name.

Reply:
^^ Still

Reply:
I don't know what to do about that.

Reply:
![image](https://user-images.githubusercontent.com/43006711/54889970-1922fa00-4e9f-11e9-9174-b2630c748e5c.png)

Reply:
tile: main tile
t_dm: tile + delta mid
t_dm_ds: tile + delta mid + delta side
t_dm_2ds: tile + delta mid + 2 times delta side
t_2dm: tile + two times delta mid
t_2dm_ds: tile + two times delta mid + delta side
t_3dm: tile + three times delta mid
t_3dm_ds: tile + three times delta mid + delta side

Reply:
Just removed all those variables, and only kept one which is reused multiple times to compute the offsets. I called it `tile_offset`.